# How far are we from personalized gene expression prediction using sequence-to-expression deep neural networks?

This repository contains the data and the scripts that were used to assess Enformer's [[1]](#1) ability to predict differential gene expression across 839 individuals [[2]](#2).

To reproduce our analysis that is shown in the paper [[2]](#2), please follow the examplary commands in Fig1.sh, Fig2.sh, and FigS.sh. 

## Abstract

_Deep learning (DL) methods accurately predict various functional properties from genomic DNA, including gene expression, promising to serve as an important tool in interpreting the full spectrum of genetic variations in personal genomes. However, systematic out-of-sample benchmarking is needed to assess the gap in their utility as personalized DNA interpreters. Using paired Whole Genome Sequencing and gene expression data we evaluate DL sequence-to-expression models, identifying their critical failure to make correct predictions on a substantial number of genomic loci, highlighting the limits of the current model training paradigm._

## References
<a id="1">[1]</a> 
Avsec, Å½., Agarwal, V., Visentin, D., Ledsam, J.R., Grabska-Barwinska, A., Taylor, K.R., Assael, Y., Jumper, J., Kohli, P. and Kelley, D.R., 2021. Effective gene expression prediction from sequence by integrating long-range interactions. [Nature methods](https://www.nature.com/articles/s41592-021-01252-x), 18(10), pp.1196-1203.

<a id="2">[2]</a>
Sasse, A.*, Ng, B.*, Spiro, A.*, Tasaki, S., Bennett, D., Gaiteri, C., De Jager, P.L., Chikina, M., and Mostafavi, S., 2023. How far are we from personalized gene expression prediction using sequence-to-expression deep neural networks? [bioRxiv](https://doi.org/10.1101/2023.03.16.532969), https://doi.org/10.1101/2023.03.16.532969, * These authors contributed equally
